### PR TITLE
Implement Form Data requests

### LINKED ISSUE #6046: Multipart file uploads are missing the boundary
## Environment
Run the following in your terminal and copy the results here.
1. `npx react-native --version`: `3.2.1`
2. `npx react-native run-windows --info`: 
```System:
    OS: Windows 10 10.0.18363
    CPU: (2) x64 Intel(R) Core(TM) i5-7600K CPU @ 3.80GHz
    Memory: 6.31 GB / 11.73 GB
  Binaries:
    Node: 12.16.3 - C:\Program Files\nodejs\node.EXE
    npm: 6.14.4 - C:\Program Files\nodejs\npm.CMD
  npmPackages:
    @react-native-community/cli: ^3.0.0 => 3.2.1
    react: 16.9.0 => 16.9.0
    react-native: 0.61.5 => 0.61.5
    react-native-windows: ^0.61.9 => 0.61.14
    rnpm-plugin-windows: ^0.4.1 => 0.4.1
  Installed UWP SDKs:
    10.0.18362.0
```
3. `reg query "HKLM\SOFTWARE\Microsoft\Windows\CurrentVersion\AppModelUnlock"`
```

HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\AppModelUnlock
    AllowDevelopmentWithoutDevLicense    REG_DWORD    0x1
    AllowAllTrustedApps    REG_DWORD    0x1
```

## Steps To Reproduce
Provide a detailed list of steps that reproduce the issue.

1. File upload using fetch
2. Fetch params :  request.headers.append("Content-Type", "multipart/form-data");

## Expected Results
File uploads
Instead server was giving a 400 response

## Snack, code example, screenshot, or link to a repository:
Please provide a Snack (https://snack.expo.io/), a link to a repository on GitHub, or provide a minimal code example that reproduces the problem.
You may provide a screenshot of the application if you think it is relevant to your bug report.
Here are some tips for providing a minimal example: https://stackoverflow.com/help/mcve

I found the attached patch fixed the issue for me

```
diff --git a/node_modules/react-native-windows/ReactUWP/Modules/NetworkingModule.cpp b/node_modules/react-native-windows/ReactUWP/Modules/NetworkingModule.cpp
index 7dde77f..9aadefc 100644
--- a/node_modules/react-native-windows/ReactUWP/Modules/NetworkingModule.cpp
+++ b/node_modules/react-native-windows/ReactUWP/Modules/NetworkingModule.cpp
@@ -300,7 +300,8 @@ void NetworkingModule::NetworkingHelper::SendRequest(
       } else if (!bodyData["formData"].empty()) {
         auto formData = bodyData["formData"];
         winrt::Windows::Web::Http::HttpMultipartFormDataContent multiPartContent;
-
+        // reset the content type as we'll need the boundary
+        contentType = nullptr;
         for (auto &formDataPart : formData) {
           if (!formDataPart["string"].empty()) {
             winrt::Windows::Web::Http::HttpStringContent multipartStringValue{
```

### LINKED ISSUE #8547: Multipart file upload header problem
## Environment
Run the following in your terminal and copy the results here.
1. `npx react-native --version`: 5.0.1
2. `npx react-native info`: 

```
System:
    OS: Windows 10 10.0.19042
    CPU: (24) x64 Intel(R) Xeon(R) CPU E5-2620 0 @ 2.00GHz
    Memory: 47.09 GB / 63.96 GB
  Binaries:
    Node: 14.17.5 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.5 - C:\Program Files (x86)\Yarn\bin\yarn.CMD
    npm: 6.14.14 - C:\Program Files\nodejs\npm.CMD
    Watchman: Not Found
  SDKs:
    Android SDK: Not Found
    Windows SDK:
      AllowDevelopmentWithoutDevLicense: Enabled
      AllowAllTrustedApps: Enabled
      Versions: 10.0.18362.0, 10.0.19041.0
  IDEs:
    Android Studio: Version     2020.3.0.0 AI-203.7717.56.2031.7621141
    Visual Studio: 16.11.31624.102 (Visual Studio Community 2019)
  Languages:
    Java: 1.8.0_171 - C:\Program Files\Java\jdk1.8.0_171\bin\javac.EXE
  npmPackages:
    @react-native-community/cli: Not Found
    react: 17.0.1 => 17.0.1
    react-native: 0.64.0 => 0.64.0
    react-native-windows: 0.64.16 => 0.64.16
  npmGlobalPackages:
    *react-native*: Not Found
```

What SDK version are you building for? Choose from 10.0.15063, 10.0.16299, 10.0.19041, etc.
- Target Platform Version(s): 10.0.18362.0, 10.0.19041.0
What device(s) are you targeting? Choose any from Desktop, Xbox, Hololens)
- Target Device(s): Desktop
Which version of Visual Studio are you using? Specify Visual Studio 2017 or Visual Studio 2019
- Visual Studio Version: Microsoft Visual Studio Community 2019, Version 16.11.2
Which build configuration are you running? Choose from Debug, DebugBundle, Release, ReleaseBundle 
- Build Configuration: Debug x64, 

## Steps To Reproduce
Provide a detailed list of steps that reproduce the issue.

1. Http post request with multipart form with code example below.
2. When executed code exception is thrown in `myApp\node_modules\react-native-windows\Shared\Modules\NetworkingModule.cpp` in function `AttachMultipartHeaders` line 247 in the second iteration of for loop     https://github.com/microsoft/react-native-windows/blob/3d7d5ff5c79390c38f8f0dcad70d15ae64c367b1/vnext/Shared/Modules/NetworkingModule.cpp#L247

## Expected Results
The post HTTP request with multipart form is sent. This issue is similar to https://github.com/microsoft/react-native-windows/issues/6046 but different place. JavaScript code is working in React Native for iOS and Android without any problem.

## Snack, code example, screenshot, or link to a repository:
```
        const formData = new FormData()

        formData.append("user", {
            string: JSON.stringify({"name": "Marcin", "family": "Nowak"}),
            type: "application/json",
        })

        const config = {
            headers: {
                "Content-Type": "multipart/form-data",
            }
        }


        const response = await fetch(
            'https://example.com/api/user/b52a6886-d02f-4121-95e3-48e3c7420081',
            {
              body: formData,
              method: "POST",
              headers: {
                'Content-Type': 'multipart/form-data',
                'Authorization': 'Bearer ' + "123"
              }
            }
          )
```

### LINKED ISSUE #9535: HTTP form data support
### Summary

Current HTTP module implementation ignores form data to be sent as part of a request.

This is due to dropping `folly` APIs from `IHttpResource` so that the interface is forward-compatible with TurboModules.

This creates the need for conversion logic from a `folly::dynamic` object containing the form data to an either ABI safe or at least Standard Library-based object type.

### Motivation

Form data body types are supported in both Android and Apple variants of React Native.

### Basic Example

_No response_

### Open Questions

_No response_

### PR DESCRIPTION
**Pull request by JunielKatarn**
## Description
Allows handling of request with enctype `multipart/form-data`.

See https://developer.mozilla.org/en-US/docs/Learn/Forms/Sending_and_retrieving_form_data

### Type of Change
- Bug fix (non-breaking change which fixes an issue)

### Why
The HTTP module is currently ignoring form-data requests.
This is a regression from the previous monolithic implementation (`NetworkingModule`).

Resolves #6046
Resolves #8547
Resolves #9535

### What
Handle request data incoming from the JS layer identified as `formData` as follows:
- Iterate each part of the form data.
  - If the part is `string`, append its headers and content to the request.
  - If the part is `uri`, append its headers and the contents of the pointed file to the request.

### Testing
- Enabled `XHRExample` for manual testing using the Playground app.

### Notes
- Reverts #10689.\
  Unrelated change broke local and CI builds.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/react-native-windows/pull/10797)

### CODE REVIEW

File: packages/e2e-test-app/windows/RNTesterApp/packages.lock.json [hunk 1]
```diff
@@ -72,8 +72,8 @@
       },
       "ReactNative.Hermes.Windows": {
         "type": "Transitive",
-        "resolved": "0.70.0",
-        "contentHash": "nh9F1wAhkndqfHG/hrS+ezesunlGhCr32pASEEAkQenn8Ou4dUwjmt0c7aT3pONl9itZ1bF9GWbpXJ7t6AsS+Q=="
+        "resolved": "0.0.0-2209.12005-9f3ce7e5",
```
Comment (aeulitz) [inline packages/e2e-test-app/windows/RNTesterApp/packages.lock.json:75]: > "0.0.0-2209.12005-9f3ce7e5", [](http://example.com/codeflow?start=19&length=29)

OOC, assuming this is not related to your change, what brings about modifications like this?

Comment (JunielKatarn) [reply packages/e2e-test-app/windows/RNTesterApp/packages.lock.json:75]: Building those packages locally.
The CI builds don't make these updates, which are VS version-related (CI was recently updated to use version 2022).

File: vnext/Shared/Networking/WinRTHttpResource.cpp [hunk 2]
```diff
@@ -146,9 +155,28 @@ IAsyncOperation<HttpRequestMessage> WinRTHttpResource::CreateRequest(
...
-    } else if (!data["form"].empty()) {
-      // #9535 - HTTP form data support
-      // winrt::Windows::Web::Http::HttpMultipartFormDataContent()
+    } else if (!data["formData"].empty()) {
+      winrt::Windows::Web::Http::HttpMultipartFormDataContent multiPartContent;
+      auto formData = data["formData"];
+
+      for (auto &formDataPart : formData) {
+        IHttpContent formContent{nullptr};
+        if (!formDataPart["string"].empty()) {
```
Comment (aeulitz) [inline vnext/Shared/Networking/WinRTHttpResource.cpp:164]: >formDataPart["string"] [](http://example.com/codeflow?start=13&length=22)

formData's 'auto' decl makes it difficult to understand what the actual type is, but if the index operator here has the same semantics as that of std::unordered_map (which is to insert one of there's no value with the respective key), would it be better to call 'at' or 'contains' here?

File: vnext/Shared/Networking/WinRTHttpResource.cpp [hunk 3]
```diff
@@ -146,9 +155,28 @@ IAsyncOperation<HttpRequestMessage> WinRTHttpResource::CreateRequest(
...
       auto stream = co_await file.OpenReadAsync();
       content = HttpStreamContent{std::move(stream)};
-    } else if (!data["form"].empty()) {
-      // #9535 - HTTP form data support
-      // winrt::Windows::Web::Http::HttpMultipartFormDataContent()
+    } else if (!data["formData"].empty()) {
+      winrt::Windows::Web::Http::HttpMultipartFormDataContent multiPartContent;
+      auto formData = data["formData"];
+
+      for (auto &formDataPart : formData) {
```
Comment (aeulitz) [inline vnext/Shared/Networking/WinRTHttpResource.cpp:162]: >auto & [](http://example.com/codeflow?start=11&length=6)

Can it be 'auto const&'?

File: packages/@react-native-windows/tester/src/js/utils/RNTesterList.windows.js [hunk 4]
```diff
@@ -12,6 +12,10 @@ import type {RNTesterModuleInfo} from '../types/RNTesterTypes';
 import ReactNativeFeatureFlags from 'react-native/Libraries/ReactNative/ReactNativeFeatureFlags';
 
 const Components: Array<RNTesterModuleInfo> = [
+  {
+    key: 'XHRExample',
+    module: require('../examples/XHR/XHRExample'),
+  },
```
Comment (aeulitz) [inline packages/@react-native-windows/tester/src/js/utils/RNTesterList.windows.js:18]: Does the addition of this array element amount to a test, or is actual test code needed to access/exercise this element?

File: vnext/Shared/Networking/WinRTHttpResource.cpp [hunk 5]
```diff
@@ -146,9 +155,28 @@ IAsyncOperation<HttpRequestMessage> WinRTHttpResource::CreateRequest(
...
+        IHttpContent formContent{nullptr};
+        if (!formDataPart["string"].empty()) {
+          formContent = HttpStringContent{to_hstring(formDataPart["string"].asString())};
+        } else if (!formDataPart["uri"].empty()) {
```
Comment (aeulitz) [inline vnext/Shared/Networking/WinRTHttpResource.cpp:166]: >formDataPart["uri"] [](http://example.com/codeflow?start=20&length=19)

are formDataPart["string"] and formDataPart["uri"] mutually exclusive?

File: vnext/Shared/Networking/WinRTHttpResource.cpp (hunk 2 above)
Comment (JunielKatarn) [reply vnext/Shared/Networking/WinRTHttpResource.cpp:164]: This object is of type `folly::dynamic`.
It can work as an array or as an unordered map.

File: packages/@react-native-windows/tester/src/js/utils/RNTesterList.windows.js (hunk 4 above)
Comment (JunielKatarn) [reply packages/@react-native-windows/tester/src/js/utils/RNTesterList.windows.js:18]: It enables existing test code not used by default:
image

**Review by aeulitz — APPROVED** [review]
:shipit:

File: vnext/Shared/Networking/WinRTHttpResource.cpp (hunk 5 above)
Comment (JunielKatarn) [reply vnext/Shared/Networking/WinRTHttpResource.cpp:166]: Yes. A single HTTP request payload can be of types:
- `string`
- `base64` (binary)
- `uri` (file content reference)
- `formData`
  - Arbitrary sequence of `string`, `base64`, or `uri`.

File: packages/@react-native/tester/js/examples/XHR/XHRExampleFetch.js [hunk 6]
```diff
@@ -28,10 +28,104 @@ class XHRExampleFetch extends React.Component<any, any> {
   }
 
   submit(uri: string) {
-    fetch(uri)
+    // fetch(uri)
```
Comment (JunielKatarn) [inline packages/@react-native/tester/js/examples/XHR/XHRExampleFetch.js:31]: Will revert / move a new example before merging.

File: vnext/Shared/Networking/WinRTHttpResource.cpp (hunk 3 above)
Comment (JunielKatarn) [reply vnext/Shared/Networking/WinRTHttpResource.cpp]: Turns out no.

Making it `const` would throw an exception for non-existing object keys.
Not sure why, but will keep as `auto&`, same as `data`.
